loaders/src/glTF/2.0/Extensions/EXT_meshopt_compression.ts

```diff
@@ -0,0 +1,66 @@
+import { Nullable } from "babylonjs/types";
+import { IGLTFLoaderExtension } from "../glTFLoaderExtension";
+import { GLTFLoader } from "../glTFLoader";
+import { IBufferView } from "../glTFLoaderInterfaces";
+
+import { MeshoptDecoder } from "./meshopt_decoder.module.js";
+
+const NAME = "EXT_meshopt_compression";
+
+/**
+ * [Specification](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression)
+ *
+ * This extension uses a WebAssembly decoder module from https://github.com/zeux/meshoptimizer/tree/master/js
+ */
+export class EXT_meshopt_compression implements IGLTFLoaderExtension {
+    /**
+     * The name of this extension.
+     */
+    public readonly name = NAME;
+
+    /**
+     * Defines whether this extension is enabled.
+     */
+    public enabled: boolean;
+
+    private _loader: GLTFLoader;
+
+    /** @hidden */
+    constructor(loader: GLTFLoader) {
+        this.enabled = loader.isExtensionUsed(NAME) && MeshoptDecoder.supported;
+        this._loader = loader;
+    }
+
+    /** @hidden */
+    public dispose() {
+    }
+
+    public loadBufferViewAsync(context: string, bufferView: IBufferView): Nullable<Promise<ArrayBufferView>> {
+        if (bufferView.extensions && bufferView.extensions[this.name]) {
+            var extensionDef = bufferView.extensions[this.name];
+            if (extensionDef._decoded) {
+                return extensionDef._decoded;
+            }
+
+            var view = this._loader.loadBufferViewAsync(context, extensionDef);
+            var decoder = MeshoptDecoder;
+
+            extensionDef._decoded = Promise.all([view, decoder.ready]).then(function (res) {
+                var source = res[0] as Uint8Array;
+                var count = extensionDef.count;
+                var stride = extensionDef.byteStride;
+                var result = new Uint8Array(new ArrayBuffer(count * stride));
+
+                decoder.decodeGltfBuffer(result, count, stride, source, extensionDef.mode, extensionDef.filter);
```

Comment:
Is this synchronous? How long does this take? Should we use web workers?

Reply:
This is synchronous; the decoders work at est. 1GB/s on a modern desktop CPU so should take tens of ms for pretty large meshes. It's possible to implement this as a web worker but given the speed it doesn't seem necessary for now at least, and it would make things a bit more complicated since you'd need to load the .js file into each worker and distribute the work.

Breakdown for Alien model in Playground example, where it takes ~1.7ms total to run the decoder, which is ~15% of the total parse time (the actual loading time is much larger here, as Mesh._refreshBoundingInfo takes ~100ms at some point later, and there's some other processing that also takes more time than raw glTF scene loading...

![image](https://user-images.githubusercontent.com/1106629/99134843-6b64e180-25d4-11eb-8927-e3831221b8ce.png)

Reply:
Yeah, there is certainly overhead when using workers. Any idea what the perf is like for mobile? Either way, it's okay for this to be synchronous for now as long as the contract is async (which it is), so we're good.

Reply:
Depends on the mobile device and whether Wasm SIMD is supported; I haven't done Wasm benchmarks on mobile. In general I would probably expect the rest of JS processing to be the bottleneck here (which is different with Draco), since the performance of Wasm code should scale according to the device power (e.g. it can easily be 100MB/s on a slow mobile CPU, but then I'd expect all the JS work to also be hit proportionally).

And yeah it's definitely possible to move this to a web worker right now, it's just that based on the testing I've done so far it doesn't seem to be necessary - if need arises, it should be straightforward to do.